dsgrid/project.py

```diff
@@ -125,8 +164,80 @@ def unload_dataset(self, dataset_id):
+            dataset = self.get_dataset(dataset_id)
+            dfs.append(dataset.get_dataframe(context, self._config))
+            logger.info("Finished processing query for dataset_id=%s", dataset_id)
+
+        if not dfs:
+            logger.warning("No data matched %s", context.name)
+            return None
+
+        # All dataset columns need to be in the same order.
+        context.consolidate_dataset_metadata()
+        # TODO: Change this when we support long format.
+        pivoted_columns = sorted(context.get_pivoted_columns())
+        dim_columns = context.get_all_dimension_query_names()
+        time_columns = context.get_dimension_query_names(DimensionType.TIME)
+        dim_columns -= time_columns
+        for col in dim_columns:
+            dimension_type = self._config.get_dimension(col).model.dimension_type
+            if dimension_type == context.get_pivoted_dimension_type():
+                dim_columns.remove(col)
+                break
+        dim_columns = sorted(dim_columns)
+        time_columns = sorted(time_columns)
+        expected_columns = time_columns + pivoted_columns + dim_columns
+        for dataset_id, i in zip(context.model.project.dataset_ids, range(len(dfs))):
+            remaining = sorted(set(dfs[i].columns).difference(expected_columns))
+            final_columns = time_columns + dim_columns + pivoted_columns + remaining
+            missing = context.get_pivoted_columns(dataset_id=dataset_id).difference(dfs[i].columns)
+            for column in missing:
+                dfs[i] = dfs[i].withColumn(column, F.lit(None).cast(DoubleType()))
+            dfs[i] = dfs[i].select(*final_columns)
+
+        main_df = dfs[0]
+        if len(dfs) > 1:
+            for df in dfs[1:]:
+                if df.columns != main_df.columns:
+                    raise Exception(
+                        f"BUG: dataset columns must match. main={main_df.columns} new={df.columns}"
+                    )
+                main_df = main_df.union(df)
+
+        return main_df
+
+    def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
+        record_ids = {}
+        base_dim_to_query_name_mapping = self._config.get_base_dimension_to_query_name_mapping()
+
+        for dim_filter in context.model.project.dimension_filters:
+            dim_type = dim_filter.dimension_type
+            query_name = dim_filter.dimension_query_name
+            df = self._config.get_dimension_records(query_name)
+            df = dim_filter.apply_filter(df).select("id")
+            if query_name != base_dim_to_query_name_mapping[dim_type]:
+                mapping_records = self._config.get_base_to_supplemental_mapping_records(query_name)
+                df = (
+                    mapping_records.join(df, on=mapping_records.to_id == df.id)
+                    .selectExpr("from_id AS id")
+                    .distinct()
+                )
+            if dim_type in record_ids:
```

Comment:
I thought it's allowing supplemental dimensions here too? Doesn't that create more than one instance of a dimension type, potentially?

Reply:
Line 239 allows multiple instances of one dimension type: base and supplemental.